feature_extractor/feature_extractor.py

```diff
@@ -101,28 +88,107 @@ def __call__(self, obs):
       'enemy_mask': self.entity_helper.enemy_mask,
 
       'game': game,
-      'legal': legal_moves,
+      'legal': self._make_legal_moves(obs),
       'prev_act': self.game_state.previous_actions(),
       'reset': np.array([self.game_state.curr_step == 0]),  # for resetting RNN hidden,
     }
 
     return state
 
+  def _make_legal_moves(self, obs):
+    legal_moves = {
+      action: np.zeros((self.team_size, dim), dtype=np.float32)
+        for action, dim in ModelArchitecture.ACTION_NUM_DIM.items()
+    }
+
+    # move
+    legal_moves["10_move"] = self.map_helper.legal_moves(obs)
+
+    # attack
+    legal_moves["21_style"] = np.ones((self.team_size,
+                                       ModelArchitecture.ACTION_NUM_DIM["21_style"]))
+    legal_moves["22_target"] = self.entity_helper.legal_target()
+
+    # use, destroy
+    if self._config.ITEM_SYSTEM_ENABLED:
+      legal_moves["30_use"] = self.item_helper.legal_inventory(obs, nmmo.action.Use)
+      legal_moves["40_destroy"] = self.item_helper.legal_inventory(obs, nmmo.action.Destroy)
+
+    if self._config.EXCHANGE_SYSTEM_ENABLED:
+      legal_moves["50_sell"] = self.item_helper.legal_inventory(obs, nmmo.action.Sell)
+      # legal_moves["60_buy"] = self.market_helper.legal_buy(obs)
+
+    # TODO: give, give-gold
+    #   give-gold and sell should benefit from a continuous price policy head
+    #   rather than a discrete action head (e.g., one-hot encoding out of 100 discrete prices)
+
+    return legal_moves
+
+  # pylint: disable=unsubscriptable-object
   def translate_actions(self, actions):
+    # save actions to game_state.prev_atns, assuming this fn is called at the step
+    self.game_state.prev_atns = actions
+
+    key_to_action = {
+      "30_use": nmmo.action.Use,
```

Comment:
prefixing move with a sort key feels hacky, if we do want to have explicit sort keys, could you create an action_to_sort_key map and use that instead?

Reply:
removed the hacky keys here, and in the `model.py`: `ModelArchitecture.ACTION_NUM_DIM`